bpf/proxy/service_type_change_test.go

```diff
@@ -0,0 +1,362 @@
+// Copyright (c) 2020 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy_test
+
+import (
+	"net"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes/fake"
+
+	"github.com/projectcalico/felix/bpf/conntrack"
+	"github.com/projectcalico/felix/bpf/mock"
+	"github.com/projectcalico/felix/bpf/nat"
+	proxy "github.com/projectcalico/felix/bpf/proxy"
+	"github.com/projectcalico/felix/ip"
+)
+
+var _ = Describe("BPF service type change", func() {
+
+	clusterIP := net.IPv4(10, 1, 0, 1)
+	extIP := net.IPv4(20, 1, 0, 1)
+	port := uint16(1234)
+	proto := v1.ProtocolTCP
+	extIPstr := "20.1.0.1"
+	npPort := int32(30333)
+	testSvc := &v1.Service{
+		TypeMeta:   typeMetaV1("Service"),
+		ObjectMeta: objectMeataV1("testService"),
+		Spec: v1.ServiceSpec{
+			ClusterIP: "10.1.0.1",
+			Type:      v1.ServiceTypeClusterIP,
+			Selector: map[string]string{
+				"app": "test",
+			},
+			Ports: []v1.ServicePort{
+				{
+					Protocol: v1.ProtocolTCP,
+					Port:     int32(port),
+				},
+			},
+		},
+	}
+
+	testSvcEps := &v1.Endpoints{
+		TypeMeta:   typeMetaV1("Endpoints"),
+		ObjectMeta: objectMeataV1("testService"),
+		Subsets: []v1.EndpointSubset{
+			{
+				Addresses: []v1.EndpointAddress{
+					{
+						IP: "10.1.2.1",
+					},
+					{
+						IP: "10.1.2.2",
+					},
+				},
+				Ports: []v1.EndpointPort{
+					{
+						Port: 1234,
+						Name: "1234",
+					},
+				},
+			},
+		},
+	}
+	k8s := fake.NewSimpleClientset(testSvc, testSvcEps)
+
+	initIP := net.IPv4(1, 1, 1, 1)
+
+	front := newMockNATMap()
+	back := newMockNATBackendMap()
+	aff := newMockAffinityMap()
+	ct := mock.NewMockMap(conntrack.MapParams)
+	p, _ := proxy.StartKubeProxy(k8s, "test-node", front, back, aff, ct, proxy.WithImmediateSync())
+	p.OnHostIPsUpdate([]net.IP{initIP})
+
+	key_clusterIP := nat.NewNATKey(clusterIP, port, proxy.ProtoV1ToIntPanic(proto))
```

Comment:
The 'C'ish naming convention is unGOish and the only true way ;-) is `camelCase`

```suggestion
	keyClusterIP := nat.NewNATKey(clusterIP, port, proxy.ProtoV1ToIntPanic(proto))
```